kernel/src/hil/digest.rs

```diff
@@ -0,0 +1,48 @@
+//! Interface for Digest
+
+use crate::returncode::ReturnCode;
+
+/// Implement this trait and use `set_client()` in order to receive callbacks.
+pub trait Client<'a> {
+    /// On error `result` will contain device specific error codes
+    fn hash_done(&'a self, result: Result<(), u32>);
+}
+
+/// Computes a digest (cryptographic hash) over data
+pub trait Digest<'a> {
+    /// Set the client instance which will receive `hash_done()` callbacks.
+    /// This callback is called when a digest is computed.
+    /// The callback should follow the `Client` `hash_done` callback.
+    fn set_client(&'a self, client: &'a dyn Client<'a>);
+
+    /// Add data to the digest block. This is the data that will be used
+    /// for the hash function.
+    /// Returns the number of bytes written on success
+    /// The data can be fed to the underlying hardware block either syncronously
+    /// or asyncronously. This is up to the hardware implementation and is
+    /// transperent to the user.
+    fn add_data(&self, data: &[u8]) -> Result<usize, ReturnCode>;
```

Comment:
Up to my knowledge it's not possible to transparently abstract over (a)synchronicity without busy-waiting. Either the component does it directly, or you'll have to wait - if this were transparent that would have to happen in the same call stack.

Therefore, a call to `add_data` in my opinion should incur a callback. If for efficiency reasons you'd like to avoid going down to the scheduler again, you could also return a boolean indicating whether a callback will be fired (asychronous operation) and hence the client would have to wait on that.

Reply:
Agreed! I'll add a callback.

Reply:
You can't always issue a callback unless you use deferred procedure calls. This adds a lot of complexity. The general programming idiom we use in Tock is to return a `(ReturnCode, Option<&[u8]>)` so you don't lose the reference to the buffer. E.g., 

https://github.com/tock/tock/blob/b134ba078c3d80a511dbd42aa05d0ff908bdd594/kernel/src/hil/uart.rs#L109

https://github.com/tock/tock/blob/b134ba078c3d80a511dbd42aa05d0ff908bdd594/kernel/src/hil/radio.rs#L103

Not all HILs have adopted this approach yet.

Reply:
@phil-levis I did not know about this approach. I would have used a deferred call here, but this seems nice as it does not unnecessarily go down to scheduler. Nonetheless, there needs to be the possibility of a callback if it were an asynchronous operation (which I guess in this instance should be the case quite often), this captures both these two options.

Reply:
Done!